### DISCUSSION TITLE
Nuclei sets HTTP Host header in raw requests

### DISCUSSION THREAD

**Discussion by tlill** [projectdiscovery/nuclei#2899] [Q&A] [answered]
### Current Behavior:
I wanted to create a template for the issue described here: https://securitytutorials.co.uk/http-header-internal-ip-disclosure/ where IIS discloses the internal IP in the Location HTTP-Header, if no Host header is set in the request and HTTP/1.0 is used. 
For that, I used the following raw HTTP request in my template:

```
[...]
requests:
  - raw:
      - |+
        GET / HTTP/1.0
        
        
    matchers-condition: and
[...]
```

But the template didn't got triggered. So after intercepting the request in Burpsuite I saw, that the request issued by nuclei was the following:
```
GET / HTTP/1.1
Host: [redacted hostname]
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:107.0) Gecko/20100101 Firefox/107.0
Connection: close
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.5


```
Here, the Host HTTP-Header is set (and some others too). This makes it impossible to create a template for this finding, as I don't know how to 'unset' or remove the Host header in nuclei.

### Expected Behavior:
If no Host header is set in the raw request, Nuclei should not set a Host header.

### Steps To Reproduce:
Use the following Template:
```
id: IIS-internal-ip-disclosure

info:
  name: Template Name
  author: T. Lill
  severity: info
  description: IIS discloses it's internal IP if no host header is set in the request and HTTP/1.0 is used.
  reference:
    - https://securitytutorials.co.uk/http-header-internal-ip-disclosure/
    - https://support.microsoft.com/en-us/topic/fix-the-internal-ip-address-of-an-iis-7-0-server-is-revealed-if-an-http-request-that-does-not-have-a-host-header-or-has-a-null-host-header-is-sent-to-the-server-c493e9bc-dfd3-0d9b-941c-b2d93a957d9e
    - https://www.cyberis.com/article/microsoft-exchange-client-access-server-information-disclosure
  tags: tags

requests:
  - raw:
      - |+
        GET / HTTP/1.0
        
        
    matchers-condition: and
    matchers:
      - type: regex
        part: header
        regex:
          - "[0-9]+.[0-9]+.[0-9]+.[0-9]+"
      - type: status
        status:
          - 301
    extractors:
      - type: regex
        name: ip
        part: header
        internal: false
        regex:
          - "Location: (.*)"
```
and run it with `nuclei -v -t IIS-internal-ip-disclosure.yaml -u <URL to IIS>`.
If the server is vulnerable, it will disclose it's internal IP in the location header of the HTTP 301 response.

**Comment by ehsandeep** [comment, marked as answer]
@tlill there is a template for it already - https://github.com/projectdiscovery/nuclei-templates/blob/master/misconfiguration/iis-internal-ip-disclosure.yaml

For malformed HTTP requests, you need to use `unsafe: true` attribute of the template, here is the documentation for the same - https://nuclei.projectdiscovery.io/templating-guide/protocols/http/#unsafe-http-requests
